nemo/collections/asr/parts/k2/w_transducer.py

```diff
@@ -0,0 +1,337 @@
+
+import torch
+import torch.nn.functional as F
+
+from nemo.collections.asr.parts.k2.graph_transducer import GraphRnntLoss, force_float32_context
+from nemo.core.utils.k2_guard import k2
+from nemo.utils.enum import PrettyStrEnum
+
+
+class GraphWTransducerLoss(GraphRnntLoss):
+    """
+    W-Transducer loss: RNN-T loss modification for training RNN-T model for the case
+    when some text at the beginning/end of the utterance is missing.
+    The resulting model behaves like the RNN-T model (no modification for decoding is required).
+    For details see "Powerful and Extensible WFST Framework for RNN-Transducer Losses" paper
+        https://ieeexplore.ieee.org/document/10096679
+    """
+
+    class LastBlankMode(PrettyStrEnum):
+        ALLOW_IGNORE = "allow_ignore"
+        FORCE_LAST = "force_last"
+
+    def __init__(
+        self,
+        blank: int,
+        eps_weight: float = 0.0,
+        last_blank_mode: Union[LastBlankMode, str] = LastBlankMode.FORCE_LAST,
+        use_grid_implementation=True,
+        connect_composed=False,
+        double_scores=False,
+        cast_to_float32=False,
+    ):
+        """
+        Init method
+
+        Args:
+            blank: blank label index
+            eps_weight: weight of epsilon transitions, 0 means no penalty (default)
+            last_blank_mode: allow to skip last blank in the prediction (default) or force it
+            use_grid_implementation: Whether to use the grid implementation (Grid-Transducer).
+            connect_composed: Connect graph after composing unit and temporal schemas
+                (only for Compose-Transducer). `connect` operation is slow, it is useful for visualization,
+                but not necessary for loss computation.
+            double_scores: Use calculation of loss in double precision (float64) in the lattice.
+                Does not significantly affect memory usage since the lattice is ~V/2 times smaller than the joint tensor.
+            cast_to_float32: Force cast joint tensor to float32 before log-softmax calculation.
+        """
+        super().__init__(
+            blank=blank,
+            use_grid_implementation=use_grid_implementation,
+            connect_composed=connect_composed,
+            double_scores=double_scores,
+            cast_to_float32=cast_to_float32,
+        )
+        self.eps_weight = eps_weight
+        self.last_blank_mode = self.LastBlankMode(last_blank_mode)
+
+    def get_unit_schema(self, units_tensor: torch.Tensor, vocab_size: int) -> "k2.Fsa":
+        """
+        Get unit schema (target text) graph for W-Transducer loss (Compose-Transducer).
+        Forward arcs represent text labels.
+
+        Example graph: text [1, 2], blank=0. Eps ids: 3, 4.
+
+        graph::
+
+                3:3:0                  0:0:1                  0:0:2
+              +-------+              +-------+              +-------+
+              v       |              v       |              v       |
```

Comment:
This ark should be `3:0:0` or `0:3:0`, shouldn't it? It is better not to have auxiliary indices in the final lattice, and so all those ark must become epsilon (0) in the end.

Reply:
Upd: I found the place where you fix this. Please see that comment.

Reply:
Won't fix for now